Quickstart guides refresh

john-rock opened the pull request:
## What are you changing in this pull request and why?
[Asana task](https://app.asana.com/0/1200099998847559/1203971242312314/f)

This PR applies the redesign for the quickstart guides.


Quickstart Guide Index: https://deploy-preview-3277--docs-getdbt-com.netlify.app/quickstarts

## Additional Information
[Figma file for index page](https://www.figma.com/file/IqnLty0ptODADf8IXCaF2U/docs---2022---2023-Website-Pages-%2F-Components?node-id=1319-2531&t=kfJBVaolk4s4paDL-0)

[Figma file for individual guide page](https://www.figma.com/file/IqnLty0ptODADf8IXCaF2U/docs---2022---2023-Website-Pages-%2F-Components?type=design&node-id=1303-2056&t=pr2V1MwkgZsI9ruq-0)

## Documentation
https://www.notion.so/dbtlabs/Quickstart-Guides-73c57db4b37547aaab4b7d9437e19eb1

- Each "step" in the quickstart guide is the content nested under each H2 (##) within the markdown file.
- Upon changing steps a `localStorage` value is set, similar to how we set versioning, that will allow users to return to the quickstart on the step they left off on.
- All quickstart guides are built from files nested in `/docs/quickstart`

## Questions
- What URL do we want the quickstart index page to live on? (currently `/quickstart`)
- Where do we want to add the link to the quickstart page? (currently nested in main nav Resources -> Quickstart guides )
- Are we redirecting the old quickstart pages to the new ones?

## Checklist
<!--
Uncomment if you're publishing docs for a prerelease version of dbt (delete if not applicable):
- [ ] Add versioning components, as described in [Versioning Docs](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-entire-pages)
- [ ] Add a note to the prerelease version [Migration Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/guides/migration/versions)
-->
- [x] Review the [Content style guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md) and [About versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version) so my content adheres to these guidelines.
- [x] All Quickstart guides are reviewed and approved in new layout

Adding new pages (delete if not applicable):
- [x] Add page to `website/sidebars.js`
- [x] Provide a unique filename for the new page

Removing or renaming existing pages (delete if not applicable):
- [x] Add an entry `_redirects`

john-rock committed:
begin work on quickstart index page

john-rock committed:
dont build spotlight index page

john-rock committed:
fix broken links for preview

john-rock committed:
work on modified toc

john-rock committed:
work on quide page

john-rock committed:
begin styling toc

john-rock committed:
cleanup

john-rock committed:
more tab cleanup

john-rock committed:
work on step butons

john-rock committed:
wip

john-rock committed:
fix introduction

john-rock commented:
Hey @JKarlavige this is still a work-in-progress but wanted to get your thoughts on a few things.
- This is using the same technique you did for the Spotlight pages to build the guide index page.
- Each `h2` within the guide is considered a step.
- The guides are written the same as they currently are on the site but the layout is modified (see next bullet point).
- The bulk of the heavy lifting to modify the layout is done within the [QuickstartTOC](https://github.com/dbt-labs/docs.getdbt.com/blob/311e4c7bb88d5b3f72d550b71e5963b32f10ef8f/website/src/components/quickstartTOC/index.js) component. It's not exactly ideal having so much heavy lifting being done there but due to the amount of custom layout on the individual guide page I attempted it. Would you recommend taking a different approach?

john-rock committed:
small fixes

john-rock committed:
fix state for initial steps

john-rock committed:
test snippet change

JKarlavige commented:
Dang, nice job getting this to work! The functionality seems solid, able to click around and all seems to work well with having the content in a single file.

Re: approach - The only way i can think of being able to avoid all the logic needed in the TOC component would be to create separate markdown files for each section, so each would be its own docs page. Then the file structure would look like:
```
docs
-- quickstarts
    -- bigquery
        -- step 1
        -- step 2
    -- redshift
        -- step 1
        -- etc.
```

With that setup, you might be able to utilize the built-in sidebar (modified to fit this style) to use as the TOC. Pagination is already built-in to docs pages as well so we wouldn't lose that functionality.

One other thought, not sure if the team would want the ability, but is it possible to navigate directly to a specific section via url? It would likely be possible with your setup using query params. If we set these up as standard docs pages, then the url would update for each step which could be one benefit of going that route.

Down to discuss this further, i think either route would work. We have more control using your custom setup, and would be able to use more built-in features but less control if building these as standard docs pages

john-rock committed:
add a few more quickstarts

john-rock committed:
work in progress

john-rock committed:
setup local storage for step

john-rock committed:
add local storage for active step on each guide

john-rock committed:
refactor local storage

john-rock committed:
attempt if check for local storage

john-rock committed:
wip

john-rock committed:
update intial state when there is no local val

john-rock committed:
fix check for quickstart layout

john-rock committed:
small refactor

john-rock committed:
index hero update

john-rock committed:
change url pattern

john-rock committed:
change pattern of ttc frontmatter to match the rest

john-rock committed:
revert on og quickstart

john-rock committed:
small refactor for snippets

john-rock committed:
actually add the file

john-rock committed:
Merge branch 'current' of https://github.com/dbt-labs/docs.getdbt.com into quickstart-redesign

john-rock committed:
adjust toc right padding

john-rock committed:
adjust font sizes

john-rock committed:
remove extra space

john-rock committed:
add padding to quickstart title

john-rock committed:
Merge branch 'current' into quickstart-redesign

john-rock committed:
remove non needed frontmatter

john-rock committed:
Merge branch 'quickstart-redesign' of https://github.com/dbt-labs/docs.getdbt.com into quickstart-redesign

john-rock committed:
add mobile styles for index

john-rock committed:
test mobile styles

john-rock committed:
minor edit

john-rock committed:
add dark mode styles

john-rock committed:
Merge branch 'current' into quickstart-redesign

john-rock committed:
cleanup index

john-rock committed:
trigger build

john-rock committed:
trigger build

john-rock committed:
cmon github actions...you can do it!

john-rock committed:
small error handling adjustment

john-rock committed:
update h2 selector

john-rock committed:
update hover

john-rock committed:
add link in sidebar

runleonarun commented on website/src/components/quickstartGuideList/index.js:
```diff
@@ -0,0 +1,49 @@
+import styles from './styles.module.css';
+
+const quickstartTitle = 'Quickstart guides'
+const quickstartDescription = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam.'
```
```suggestion
const quickstartDescription = 'dbt Core is a powerful open-source tool for [data transformations](https://www.getdbt.com/analytics-engineering/transformation/), and dbt Cloud is the fastest and most reliable way to deploy your dbt jobs. With the help of a sample project, learn how to quickly start using dbt and one of the most common data platforms.'
```

john-rock replied:
Hey @runleonarun, if we wanted to use a link in the hero component subheading we would need to add the `dangerouslySetInnerHTML` prop to the subheading which we should avoid using if possible to avoid [cross-site scripting (XSS)](https://en.wikipedia.org/wiki/Cross-site_scripting) attacks. We can add it if we really wanted to but wanted to call that out.

I just pushed the content suggested above without the link, let me know what you think.

runleonarun replied:
That works!

runleonarun reviewed (commented):
Hey @john-rock Just a minor suggestion for the Overview!

john-rock committed:
update index subheader

john-rock committed:
add start button style

Commit:
change headers

mirnawong1 committed:
Merge branch 'current' into quickstart-redesign

mirnawong1 commented:
hey @john-rock, this looks awesome! ok so per @runleonarun and yourself, I've made changes to the H2/H3 headers to try and consolidate some of the content for the left sidebar steps. i think I've eliminated 2-3 steps and made the following changes:

- Turned the H2 Prerequisites into H3 and nested it under the H2 Introduction
- Turned the H2 Related docs into H3 and moved it to the top so it falls under H2 Introduction
- Consolidated the deploy jobs snippet into 1 step, rather than 2 steps (Added an H2 'Deploy dbt' and turned 'Create an env and create a job' into H3. 
- However, when I made those changes - I only saw the 'Deploy dbt' change and didn't see the 'Prerequisites' and 'Related docs' content change at all. Not sure if that was just because of my local host?

Also noticed some weird sidebar behavior in the dbt core manual install, [loom here](https://www.loom.com/share/745c307567854c3ebbd8b2ec298c1b74). let me know if that's all ok, thanks so much for this great work!

john-rock committed:
update correct directory

john-rock committed:
update core guides

john-rock committed:
remove old quickstart directory

john-rock commented:
> However, when I made those changes - I only saw the 'Deploy dbt' change and didn't see the 'Prerequisites' and 'Related docs' content change at all. Not sure if that was just because of my local host?

Hey @mirnawong1 , regarding the comment above. The new quickstart files are built from the new directory `website/docs/quickstart` whereas the changes you applied are to the old quickstart directory under `website/docs/docs/quickstarts`. That was on me, I forgot to remove the old files in the previous commit but I just went ahead an applied your updates in the new directory and removed the previous quickstart directory.

I'm taking a look into the curious case of the shrinking sidebar and should have a fix in place shortly, good catch!

john-rock committed:
add files back

john-rock committed:
restore og quickstarts

john-rock committed:
restore og quickstarts

john-rock committed:
prevent shrinking sidebar

mirnawong1 commented:
hey @john-rock oh ok gotcha, completely didn't realize! sorry about that and thank you for doing the updates 🙏  i see the changes in the [deploy preview](https://deploy-preview-3277--docs-getdbt-com.netlify.app/quickstart) now 💥 

In terms of the new URL pattern, i personally think /`quickstart/codespace` is fine and cleaner but I'll defer to @runleonarun @nghi-ly on that.

john-rock committed:
prevent error when snippet is nested

mirnawong1 committed:
Merge branch 'current' into quickstart-redesign

runleonarun commented:
Hi @john-rock! 
> Actually before I remove the old files are we okay with the new URL pattern (e.g. /quickstart/codespace)? Once we remove the old files we will need to go in and update the links to point to the new location.

- [ ] Can we take quickstarts out of the Resources drop down and move it to Resources > Guides > Quickstarts? 
- [ ] I'd like also it to be the landing page when you click Resources > guides.  
- [ ] The URL that might make most sense would be:  /guides/quickstarts/codespace

john-rock committed:
add query param on active step

john-rock committed:
Merge branch 'quickstart-redesign' of https://github.com/dbt-labs/docs.getdbt.com into quickstart-redesign

john-rock committed:
fix local storage access

john-rock committed:
fix next logic

nghi-ly commented:
hi @john-rock ! thank you so much for all this work! 🙏 

it might make sense to add a product name in the URL too like https://docs.getdbt.com/docs/quickstarts/dbt-cloud/bigquery or https://docs.getdbt.com/docs/quickstarts/dbt-core/manual-install . this is for situations where we may want to create quickstarts for both products (i don't know what that may be but gives us options -- future proofing). and maybe it might help with filtering in the future in some way 🤷‍♀️

john-rock committed:
fix param null

john-rock commented:
> it might make sense to add a product name in the URL too like https://docs.getdbt.com/docs/quickstarts/dbt-cloud/bigquery or https://docs.getdbt.com/docs/quickstarts/dbt-core/manual-install . this is for situations where we may want to create quickstarts for both products (i don't know what that may be but gives us options -- future proofing). and maybe it might help with filtering in the future in some way 🤷‍♀️

Hey @nghi-ly, unfortunatley with the way the pages are steup to build currently we wouldn't bne able to seperate them out into differnet directories depedning on the product from what I understand.

What we can do currently if we have the same quickstarts for each product we can prefix the id frontamatter value to correspond to that product
E.g.
```
title: "Quickstart for dbt Core using GitHub Codespaces"
id: dbt-core-codespace
```
```
title: "Quickstart for dbt Cloud using GitHub Codespaces"
id: dbt-cloud-codespace
```

In each of the above examples the URLs would translate to `/quickstart/dbt-core-codespace` and `/quickstart/dbt-cloud-codespace` respectivly. We do plan on doing some filtering functionality in a V2 on the index page and during that time we can add a label to the cards as well to identify which product they are for.

nghi-ly commented:
@john-rock , great workaround! lgtm! thank you for the extra context, helps me understand our situation better 🙏

nghi-ly commented on website/docs/docs/quickstarts/dbt-core/codespace-qs.md:
```diff
@@ -5,15 +5,25 @@ description: "Quickly get set up to start running dbt commands in a codespace en
 - To use the dbt command-line interface (CLI), it's important that you know some basics of the terminal. In particular, you should understand `cd`, `ls` , and `pwd` to navigate through the directory structure of your computer easily.
 - You have a [GitHub account](https://github.com/join).
 
+## Related content
```
converting to H3 so it's on the same page as Prereqs

```suggestion
### Related content
```

nghi-ly commented:
i love that it remembers which step i was last at in the quickstart (when i jump around the diff guides)! maybe it's browser cache 🤔  but still a lovely experience.

john-rock commented:
> i love that it remembers which step i was last at in the quickstart (when i jump around the diff guides)! maybe it's browser cache 🤔 but still a lovely experience.

We are setting the active step on the guides via `localStorage` in the browser 🙂. On page load we are checking for either the `step=` URL query param or the `localStorage` value to determine which step to load first.

runleonarun commented:
Hey @john-rock I'm seeing the complications in what I suggested [here](https://github.com/dbt-labs/docs.getdbt.com/pull/3277#issuecomment-1551728025). We can chat about options today in the meeting.

JKarlavige commented on website/docs/docs/quickstarts/dbt-cloud/bigquery-qs.md:
❔ Will these old quickstarts remain on the site, or will these be removed and redirected to the new quickstarts?

john-rock replied:
We're redirecting but held off on deleting them until we confirm the location of the new quickstarts. We should have that sorted out after our sync today and will update this accordingly.

JKarlavige commented on website/src/components/quickstartTOC/index.js:
```diff
@@ -0,0 +1,223 @@
+  const [mounted, setMounted] = useState(false);
+  const [tocData, setTocData] = useState([]);
+  const [activeStep, setActiveStep] = useState(activeStepLocal || 1);
+  const [activeQuickstart, setActiveQuickstart] = useState(locationPath);
```
eslint stating `'setActiveQuickstart' is assigned a value but never used`. Is it possible to remove this state and use `locationPath` directly, or do we anticipate the `activeQuickstart` state changing down the road?

john-rock replied:
Good call. I don't anticipate needing that state down the road. Just nuked it in favor of `locationPath`.

JKarlavige commented on website/src/components/quickstartTOC/index.js:
```diff
@@ -0,0 +1,223 @@
+    const steps = quickstartContainer.querySelectorAll("h2");
+    const snippetContainer = document.querySelectorAll(".snippet");
+
+    // Add snippet container to its parent step
+    snippetContainer.forEach((snippet) => {
+      const parent = snippet?.parentNode;
+      while (snippet?.firstChild && parent.className) {
+        if (parent) {
+          parent.insertBefore(snippet.firstChild, snippet);
+        }
+      }
+    });
```
Is there additional work needed to handle snippets for quickstarts? Or does this loop take care of that automatically?

john-rock replied:
That loop takes care of it. The only thing I had to add to the snippet component was a class, `snippet`, in order for the `forEach` to accurately read all actual snippets.

JKarlavige commented on website/src/components/quickstartTOC/index.js:
```diff
@@ -0,0 +1,223 @@
+    setMounted(true);
+    setActiveStep(activeStepParam || parseInt(activeStepLocal) || 1);
+
+    // Wrap all h2 (steps), along with all of their direct siblings, in a div until the next h2
```
Nice work commenting on all functionality, definitely helps make everything easier to understand what's going on. Lots happening in this file, nice work structuring everything.

JKarlavige commented on website/src/components/quickstartTOC/index.js:
```diff
@@ -0,0 +1,223 @@
+    });
+
+    // Create an array of objects with the id and title of each step
+    const data = Array.from(steps).map((step, index) => ({
+      id: step.id,
+      title: step.innerText,
+      stepNumber: index + 1,
+    }));
```
🔥🔥🔥

JKarlavige commented on website/src/components/quickstartTOC/index.js:
```diff
@@ -0,0 +1,223 @@
+
+    setActiveStep(newStepIndex);
+
+    console.log('current step: ', currentStepIndex)
+    console.log('new step: ', newStepIndex)
```
Just a note to nuke consoles before take-live

JKarlavige commented on website/src/components/quickstartTOC/index.js:
```diff
@@ -0,0 +1,223 @@
+  };
+
+  // Handle TOC click
+  const handleTocClick = (e) => {
+    const stepNumber = parseInt(e.target.dataset.step);
+
+    const currentStep = document.querySelector(`.${style.stepWrapper}:not(.${style.hidden})`);
+    currentStep?.classList.add(style.hidden);
+
+    const newStep = document.querySelector(
+      `.${style.stepWrapper}[data-step='${stepNumber}']`
+    );
+    newStep?.classList.remove(style.hidden);
+
+    setActiveStep(stepNumber);
```
Would it be possible to use the `updateStep` method within the `handleTocClick` method here? It looks like the way the `currentStep` is set is different here than the `updateStep`, so not sure if that requires this to be separate.

JKarlavige replied:
Noticing one thing occurring when clicking the TOC items. If you click the number directly, it does not get the step and does not show content. It also adds `NaN` to the query param.

Loom video: https://www.loom.com/share/bd1aea6694b047e6bfb90ea03e8832e3

john-rock replied:
I'll look into both of these. Good catch with clicking the numbers in the TOC

john-rock replied:
Was able to fix both of these. There was some previous work left over in the `handleTocClick`, updated it to use `updateState` function.

Also refactors the TOC number NaN bug. Let me know if you're still seeing it on your end.

JKarlavige reviewed (changes requested):
Have one requested change regarding the TOC items when the numbers are clicked directly. I left a comment with a loom video showing the issue. That was the only thing I can find, this functions really well. 

I think you chose the right route to build this out, great work. Love that it saves the steps and that you can go to a step directly via a query param.

john-rock committed:
update quickstart name

john-rock committed:
update NaN on toc number click

john-rock committed:
add redirects

john-rock committed:
remove non needed state

john-rock committed:
update links

john-rock commented:
Hey @JKarlavige re-requesting review after applying your feedback.

@nghi-ly @runleonarun with the most recent commits I've removed the old quickstart files, updating the links, and added redirects where necessary. When you have a moment may you give this PR a review as well as review the content in each quickstart to make sure everything is looking good? Tagging @mirnawong1 also since you made some content updates during this project.

As always feel free to ask any remaining questions you may have!

mirnawong1 committed:
Merge branch 'current' into quickstart-redesign

mirnawong1 commented on website/src/components/quickstartGuideList/index.js:
```diff
@@ -0,0 +1,49 @@
+import styles from './styles.module.css';
+
+const quickstartTitle = 'Quickstarts'
+const quickstartDescription = 'dbt Core is a powerful open-source tool for data transformations and dbt Cloud is the fastest and most reliable way to deploy your dbt jobs. With the help of a sample project, learn how to quickly start using dbt and one of the most common data platforms.'
```
hey @john-rock , this looks incredible!
- i moved dbt cloud first before dbt core and also adding a longer description to make it clear what the user is going to get out of this. but feel free to cut it if it's too long. i tried adding a line break but it didn't render nicely 🙈 this is a non blocking suggestion and apologies if i added more work!

```suggestion
const quickstartDescription = 'dbt Core is a powerful open-source tool for data transformations and dbt Cloud is the fastest and most reliable way to deploy your dbt jobs. With the help of a sample project, learn how to quickly start using dbt and one of the most common data platforms.'
```

mirnawong1 replied:
also @john-rock hey @john-rock , i think is almost good to go on my end! one more non blocking suggestion:

is it possible to sort the cards so that cloud quickstart are together and then core quickstart are together? this is to make it easier for the user to select their path:

- the first row can be Cloud with BigQuery, Cloud with Databricks, Cloud with Redshfit.
- 2nd row is Cloud with Snowflake, Cloud with Starburst, Core with Codespace
- 3rd row is Core with manual install.

john-rock replied:
Hey @mirnawong1! Thanks for the feedback. 

Most of the sorting enhancements will be included in a V2 of this. Unfortunately right now I won't have the capacity to apply directly what you are suggesting above.

What I can do, and this is now available in the preview, is add a frontmatter key (`platform`) and added a conditional check to sort all quickstarts with the frontmatter `platform = 'dbt-cloud'` first following by the `dbt-core` quickstarts.

Let me know if you are okay with that solution for now.

john-rock replied:
As for the content in the hero section, right now that component is setup to only accept a string for the value of the `subheading`. Because of this we currently can not add any html as it would always be rendered as a string.

Here's a screenshot of your suggested text below. IMO, it does look a bit long but let me know what you think.

![image](https://github.com/dbt-labs/docs.getdbt.com/assets/46692803/f210d12a-5318-4a54-b68c-17ea49c8d069)

mirnawong1 replied:
hey @john-rock, i’ll remove/update my content suggestion as youre so right, it looks too long!

john-rock committed:
sort by dbt cloud first

mirnawong1 committed:
Merge branch 'current' into quickstart-redesign

mirnawong1 commented:
looks good to me @john-rock

JKarlavige reviewed (dismissed):
lgtm, great work with this! :shipit:

john-rock committed:
update sorting function

john-rock committed:
Merge branch 'quickstart-redesign' of https://github.com/dbt-labs/docs.getdbt.com into quickstart-redesign

john-rock committed:
trigger build

john-rock committed:
trigger build

john-rock committed:
fix merge conflict

JKarlavige reviewed (approved):
:shipit: :shipit: :shipit:

mirnawong1 commented:
incredible job @john-rock ! thank you so much for making these beautiful changes and i appreciate you asking for feedback ❤️ ! nice job team and looks amazing to me 😄

runleonarun reviewed (approved):
Let's ship this PR and then the Docs team can follow up with some tweaks to titles and such! @john-rock Thanks so much for your help and making this look so awesome!

john-rock committed:
Merge branch 'current' into quickstart-redesign
